tests/e2e/v3_curl_test.go

```diff
@@ -385,6 +386,107 @@ func testV3CurlMaintenanceAlarmMissiongAlarm(cx ctlCtx) {
 	}
 }
 
+func TestV3CurlMaintenanceStatus(t *testing.T) {
+	testCtl(t, testV3CurlMaintenanceStatus, withCfg(*e2e.NewConfigNoTLS()))
+}
+
+func testV3CurlMaintenanceStatus(cx ctlCtx) {
+	if err := e2e.CURLPost(cx.epc, e2e.CURLReq{
+		Endpoint: "/v3/maintenance/status",
+		Value:    "{}",
+		Expected: expect.ExpectedResponse{
+			Value:         fmt.Sprintf(`.*"version":%q,"dbSize":.*,\s*"leader":.*,\s*"raftIndex":.*,\s*"raftTerm":.*,\s*"raftAppliedIndex":.*,\s*"dbSizeInUse":.*,\s*"storageVersion":.*`, version.Version),
```

Comment:
I don't understand why you are testing json using regex? This creates a invalid dependency on order of keys.

Could you just marshall it to map and check keys?

Reply:
> I don't understand why you are testing json using regex? 

Because we don't care the values, which are also uncertain.

> This creates a invalid dependency on order of keys.

The objects are marshalled in the order that fields are declared, why it is invalid? Note I don't think we will change the order of the fields frequently.

Reply:
also read https://github.com/etcd-io/etcd/pull/16472

Reply:
> The objects are marshalled in the order that fields are declared, why it is invalid? Note I don't think we will change the order of the fields frequently.

Order of the API fields is not something we guarantee and something we should test. I expect it just unexpected side effect in the current implementation of grpc gateway. We should not write tests based on unexpected side effects.

See https://github.com/etcd-io/etcd/blob/d09b8d474a01ab4ebf4a7bd877962c1036fb2b68/tests/e2e/cmux_test.go#L140-L171

Reply:
I don't agree with https://github.com/etcd-io/etcd/pull/16472 it's unnesesery overkill that will lead to bad code patterns (aka, use regex for everything) and bloat (regex patterns gets so complicated we stop understanding them).

Reply:
I agree the solution isn't perfect. But I'd like keep it as it's.

Only the order of map fields are clearly mentioned not guaranteed. Struct fields are stable for now, although I don't see clear document on that. 

Unmarshalling the response into a struct without error doesn't mean it's a valid or expected response. Json by default ignore the unknown fields. Even the server side gives your a "{}" response, You can also unmarshal it into most of the existing PB structs in etcd. You have to double check whether the values are expected. Comparing the string directly (exact matching or regular expression matching) doesn't have such issue.

Enhancing the test framework to support regular expression matching doesn't mean we will use it for everything. Also the existing regular expressions what we used is simple enough.

Reply:
The duplicated message definitions seem also not good. https://github.com/etcd-io/etcd/blob/d09b8d474a01ab4ebf4a7bd877962c1036fb2b68/tests/e2e/cmux_test.go#L157-L168

Reply:
> Json by default ignore the unknown fields.

We don't need to do that. 

> Even the server side gives your a "{}" response, You can also unmarshal it into most of the existing PB structs in etcd. You have to double check whether the values are expected. Comparing the string directly (exact matching or regular expression matching) doesn't have such issue.
 
 You can unmarshal into `map[string]interface{}` and check keys.

Reply:
> The duplicated message definitions seem also not good.

yea, it's an overkill here.

Reply:
> You can unmarshal into map[string]interface{} and check keys.

Actually I thought about this approach as well. The reason why I didn't do it is it complicates the test cases, especially when there are nested structs.

Reply:
Updated anyway, PTAL.